### PR TITLE
Burn fees collected into invalid accounts

### PR DESCRIPTION
**Pull request by jstarry**
#### Problem
Validators can currently collect transaction and rent fees into any account type they wish. This is unnecessarily permissive and fees should only be allowed to be collected into system owned accounts that are rent exempt.

#### Summary of Changes
- Refactor: move fee distribution code into new `bank::fee_distribution` module to avoid increasing the size of `bank.rs`
- Prevent collecting transaction and rent fees into non system owned accounts
- Prevent the creation of new rent paying accounts via transaction fee collection
- Move `Bank::deposit` into `test_utils` module to discourage its usage

Feature Gate Issue: https://github.com/solana-labs/solana/issues/33888

### CODE REVIEW

File: runtime/src/bank/fee_distribution.rs [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+    //
+    // collector_id is rotated according to stake-weighted leader schedule. So the opportunity of
+    // earning transaction fees are fairly distributed by stake. And missing the opportunity
+    // (not producing a block as a leader) earns nothing. So, being online is incentivized as a
+    // form of transaction fees as well.
+    //
+    // On the other hand, rent fees are distributed under slightly different philosophy, while
+    // still being stake-weighted.
+    // Ref: distribute_rent_to_validators
+    pub(super) fn distribute_transaction_fees(&self) {
```
Comment (CriesofCarrots) [inline runtime/src/bank/fee_distribution.rs:47]: In future, please limit move commits to just the move by itself and rename methods in a separate commit.

File: runtime/src/bank.rs [hunk 2]
```diff
@@ -3805,7 +3804,7 @@ impl Bank {
             .stakes_cache
             .stakes()
             .highest_staked_node()
-            .unwrap_or_default();
```
Comment (CriesofCarrots) [inline runtime/src/bank.rs:3864]: Why this change?

File: runtime/src/bank/fee_distribution.rs [hunk 3]
```diff
@@ -0,0 +1,953 @@
...
+        // Check the result
+        match result {
+            Ok(_) => panic!("Expected an error due to lamport overflow, but received Ok"),
+            Err(e) => {
+                match e {
+                    DepositFeeError::LamportOverflow => {} // This is the expected error
+                    _ => panic!("Unexpected error: {:?}", e),
+                }
+            }
+        }
```
Comment (CriesofCarrots) [inline runtime/src/bank/fee_distribution.rs:543]:
```suggestion
        let error = bank.deposit_fees(&pubkey, deposit_amount, options).expect_err("Expected an error due to lamport overflow, but received Ok");
        assert_eq!(error, DepositFeeError::LamportOverflow);
```
This will necessitate deriving PartialEq on DepositFeeError, but seems worth it to me.

File: runtime/src/bank/fee_distribution.rs [hunk 4]
```diff
@@ -0,0 +1,953 @@
...
+        // Check the result
+        match result {
+            Ok(new_balance) => {
+                assert_eq!(
+                    new_balance,
+                    initial_balance + deposit_amount,
+                    "New balance should be the sum of the initial balance and deposit amount"
+                );
+            }
+            Err(e) => panic!("Unexpected error: {:?}", e),
```
Comment (CriesofCarrots) [inline runtime/src/bank/fee_distribution.rs:514]: Unwrapping above would print the error, wouldn't it? I feel like the match syntax doesn't gain us much here.

File: runtime/src/bank/fee_distribution.rs [hunk 5]
```diff
@@ -161,53 +212,40 @@ impl Bank {
...
-                        warn!(
-                            "Rent distribution of {rent_to_be_paid} to {pubkey} results in \
-                            invalid RentState: {recipient_post_rent_state:?}"
-                        );
-                        datapoint_warn!(
-                            "bank-rent_distribution_invalid_state",
-                            ("slot", self.slot(), i64),
-                            ("pubkey", pubkey.to_string(), String),
-                            ("rent_to_be_paid", rent_to_be_paid, i64)
-                        );
```
Comment (CriesofCarrots) [inline runtime/src/bank/fee_distribution.rs]: Just calling out that we do lose this warning and datapoint in the new `deposit_fees`. Hopefully, `prevent_rent_paying_rent_recipients()` will be active before this PR is released, so we'll no longer care about tracking that specific case so carefully.

File: runtime/src/bank/fee_distribution.rs [hunk 6]
```diff
@@ -161,53 +212,40 @@ impl Bank {
...
+                            self.capitalization.fetch_sub(rent_to_be_paid, Relaxed);
+                            error!(
+                                "Burned {} rent lamports instead of sending to {}",
+                                rent_to_be_paid, pubkey
+                            );
+                            datapoint_error!(
+                                "bank-burned_rent",
+                                ("slot", self.slot(), i64),
+                                ("num_lamports", rent_to_be_paid, i64)
+                            );
```
Comment (CriesofCarrots) [inline runtime/src/bank/fee_distribution.rs:247]: Do we maybe want to back down these, or at least the log? It's not really the kind of thing a node operator would interpret as an error.

File: runtime/src/bank/fee_distribution.rs [hunk 7]
```diff
@@ -50,24 +69,56 @@ impl Bank {
...
+                    Err(err) => {
+                        error!(
+                            "Burning {} fee instead of crediting {} due to {err}",
+                            deposit, self.collector_id
+                        );
+                        datapoint_error!(
+                            "bank-burned_fee",
+                            ("slot", self.slot(), i64),
+                            ("num_lamports", deposit, i64)
+                        );
```
Comment (CriesofCarrots) [inline runtime/src/bank/fee_distribution.rs:81]: Same question about reducing from `error`

**Review by CriesofCarrots — COMMENTED** [review]
Just a couple questions and nits.
I only commented on a couple, but I'd love to see all the test `match result`s cleaned up for readability.

File: runtime/src/bank.rs (hunk 2 above)
Comment (jstarry) [reply runtime/src/bank.rs:3864]: The default key is the system program key which is owned by the native loader and so cannot be used as a collector id

Comment (CriesofCarrots) [reply runtime/src/bank.rs:3864]: Oh right. Can you please add a comment explaining? Or should this fallback maybe be `incinerator::id()`

File: runtime/src/bank/fee_distribution.rs (hunk 6 above)
Comment (jstarry) [reply runtime/src/bank/fee_distribution.rs:247]: Yeah good idea. I think that if a validator is burning its own fees it should see an error but it won't care about other validators. Do you care if I remove the datapoint?

File: runtime/src/bank/fee_distribution.rs (hunk 5 above)
Comment (jstarry) [reply runtime/src/bank/fee_distribution.rs]: Yeah I was curious if you wanted to keep that or not, I can keep it in if you want!

Comment (CriesofCarrots) [reply runtime/src/bank/fee_distribution.rs]: @jeffwashington , @brooksprumo , I think you're the main users of this datapoint currently. Thoughts on removing it?

File: runtime/src/bank/fee_distribution.rs (hunk 6 above)
Comment (CriesofCarrots) [reply runtime/src/bank/fee_distribution.rs]: I don't know of a need for the datapoint and am personally fine with its removal, but that might be a good question for @jeffwashington or @HaoranYi

File: runtime/src/bank/fee_distribution.rs [hunk 8]
```diff
@@ -0,0 +1,893 @@
...
+                                    reward_type: RewardType::Rent,
+                                    lamports: rent_to_be_paid as i64,
+                                    post_balance,
+                                    commission: None,
+                                },
+                            ));
+                        }
+                        Err(err) => {
+                            // overflow adding lamports or resulting account is invalid
+                            self.capitalization.fetch_sub(rent_to_be_paid, Relaxed);
```
Comment (HaoranYi) [inline runtime/src/bank/fee_distribution.rs:234]: Instead of burning the fees here, do we consider another option - we distribute these rents to other staked nodes? yet another thought ...

File: runtime/src/bank/fee_distribution.rs [hunk 9]
```diff
@@ -0,0 +1,893 @@
...
+    // utilization cost.
+    //
+    // We can't use collector_id (which is rotated according to stake-weighted leader schedule)
+    // as an approximation to the ideal rent distribution to simplify and avoid this per-slot
+    // computation for the distribution (time: N log N, space: N acct. stores; N = # of
+    // validators).
+    // The reason is that rent fee doesn't need to be incentivized for throughput unlike transaction
+    // fees
+    //
+    // Ref: collect_transaction_fees
```
Comment (HaoranYi) [inline runtime/src/bank/fee_distribution.rs:137]: nit: distribute_transaction_fees.

**Review by HaoranYi — COMMENTED** [review]
Nice work.  Just a few nits and suggestions.

File: runtime/src/bank/fee_distribution.rs [hunk 10]
```diff
@@ -0,0 +1,893 @@
...
+                        burn += deposit;
+                    }
+                }
+            }
+            self.capitalization.fetch_sub(burn, Relaxed);
+        }
+    }
+
+    // Deposits fees into a specified account and if successful, returns the new balance of that account
+    fn deposit_fees(
```
[the remark this answers is not in the document]
Comment (jstarry) [reply runtime/src/bank/fee_distribution.rs:87]: We need the same checks (including rent state transition) in both cases, that's why I created this function. If we collect transaction fees into an account whose post balance is not greater than or equal to the rent exempt balance, we need to skip the fee deposit because otherwise we allow creating rent paying accounts. Make sense?

File: runtime/src/bank/fee_distribution.rs (hunk 8 above)
Comment (jstarry) [reply runtime/src/bank/fee_distribution.rs:234]: Yeah it's a good thought but I think it's not worth the effort and complexity. Transaction fees shouldn't be distributed to other staked nodes because they only go to the block producer. Rent fees could be re-distributed but they will be phased out over time once there are basically no more rent paying accounts left on the cluster.

File: runtime/src/bank.rs (hunk 2 above)
Comment (jstarry) [reply runtime/src/bank.rs:3864]: Yeah that's unfortunate about `cfg(test)` not propagating, I forgot about that. I'll add a comment and keep it as a new unique pubkey. It's not possible to set the incinerator as the collector id in normal cluster operation so I'd rather not use that here even for tests.

File: runtime/src/bank/fee_distribution.rs (hunk 6 above)
Comment (jstarry) [reply runtime/src/bank/fee_distribution.rs]: @HaoranYi is there a specific reason why you would like to track this datapoint? I can't think of any reason personally.

File: runtime/src/bank.rs (hunk 2 above)
Comment (CriesofCarrots) [reply runtime/src/bank.rs:3864]: > It's not possible to set the incinerator as the collector id in normal cluster operation

Sorry if I'm missing something obvious, but why is that?

File: runtime/src/bank/fee_distribution.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (HaoranYi) [reply runtime/src/bank/fee_distribution.rs:87]: Yeah. 
It seems that the current code still allows transaction fees to be collected into an account whose post balance is not greater than or equal to the rent exempt balance. Should we create a separate PR to fix this for current code and backport it to v1.17?

File: runtime/src/bank/fee_distribution.rs (hunk 6 above)
Comment (HaoranYi) [reply runtime/src/bank/fee_distribution.rs]: It gives us the visibility of how much fees are burned in the network. The amount of the fees burned should be very small.  To me, being able to run such a query and verify the amount of fees burned, could be a useful sanity check.

File: runtime/src/bank.rs (hunk 2 above)
Comment (jstarry) [reply runtime/src/bank.rs:3864]: Collector ID needs to sign to be set as the node id for a vote account and that node id is what's used as the collector id

File: runtime/src/bank/fee_distribution.rs (hunk 6 above)
Comment (jstarry) [reply runtime/src/bank/fee_distribution.rs]: Good point, I'll add it back in, thanks

File: runtime/src/bank/fee_distribution.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (jstarry) [reply runtime/src/bank/fee_distribution.rs:87]: I'm not sure what you mean, can you elaborate more? By the "current code" you mean the code in this PR right? And under what conditions are you saying that the code "still allows transaction fees to be collected into an account whose post balance is not greater than or equal to the rent exempt balance"?

File: runtime/src/bank/fee_distribution.rs (hunk 5 above)
Comment (jstarry) [reply runtime/src/bank/fee_distribution.rs]: The latest patch of this PR removes this particular datapoint_warn but has two datapoint_warn's that track how many lamports were burned during both tx and rent fee distribution

```rust
datapoint_warn!(
    "bank-burned_rent",
    ("slot", self.slot(), i64),
    ("num_lamports", rent_to_burn, i64)
);

datapoint_warn!(
    "bank-burned_fee",
    ("slot", self.slot(), i64),
    ("num_lamports", deposit, i64),
    ("error", err.to_string(), String),
);
```

File: runtime/src/bank/fee_distribution.rs (hunk 6 above)
Comment (jstarry) [reply runtime/src/bank/fee_distribution.rs]: I added the datapoint back in but downgraded it to "warn"

File: runtime/src/bank/fee_distribution.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (HaoranYi) [reply runtime/src/bank/fee_distribution.rs:92]: By "current code", I mean master. 

In master, we don't validate the post balance of the account that transaction fees are deposit to. If the balance of the account pre-transaction fee was zero, then after the transaction fee deposit, we introduce a "new" rent paying account, right?

The new code (i.e. deposit_fee fn) in the PR will prevent this from happening. What I mean is that we pull this piece out in a new PR to patch master separately.

File: runtime/src/bank/fee_distribution.rs [hunk 11]
```diff
@@ -230,20 +236,29 @@ impl Bank {
...
+                            // so burn lamports and track lamports burned per slot
+                            rent_to_burn = rent_to_burn.saturating_add(rent_to_be_paid);
                         }
                     }
                 }
             });
         self.rewards.write().unwrap().append(&mut rewards);
 
+        if rent_to_burn > 0 {
+            self.capitalization.fetch_sub(rent_to_burn, Relaxed);
```
Comment (CriesofCarrots) [inline runtime/src/bank/fee_distribution.rs:254]: This change to a single subtraction per loop looks fine, but since it was potentially substantive, it would have been nice if called out in its own commit.

**Review by CriesofCarrots — APPROVED** [review]
Lgtm. It just occurred to me that, since this adds a feature gate, it really should have been a SIMD first.

File: runtime/src/bank/fee_distribution.rs (hunk 5 above)
Comment (jstarry) [reply runtime/src/bank/fee_distribution.rs]: For `bank-burned_fee` we track the slot which allows us to infer who the leader was that failed to collect tx fees

File: runtime/src/bank/fee_distribution.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (jstarry) [reply runtime/src/bank/fee_distribution.rs:92]: I see, I'd rather not break up apart this PR unless we have a good reason to. I purposely setup this PR to introduce the shared `deposit_fee` function to ensure that all invalid cases are handled for each type of fee collection. Before, without a common function, we missed the rent paying edge case for collected transaction fees.

**Comment by jstarry** [comment]
> Lgtm. It just occurred to me that, since this adds a feature gate, it really should have been a SIMD first.

Good point, I created a SIMD to improve visibility and explain this change in more detail: https://github.com/solana-foundation/solana-improvement-documents/pull/85

File: runtime/src/bank/fee_distribution.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (HaoranYi) [reply runtime/src/bank/fee_distribution.rs:92]: OK. I think we might want to port this feature back to 1.17 to close those edge cases as soon as possible.

**Review by HaoranYi — APPROVED** [review]
lgtm

File: runtime/src/bank.rs [hunk 12]
```diff
@@ -3805,7 +3804,7 @@ impl Bank {
...
+            .unwrap_or_else(Pubkey::new_unique);
```
Comment (t-nelson) [inline runtime/src/bank.rs:3809]: this method should absolutely not be used outside of tests under any circumstances.  can we just fail/panic instead?

File: runtime/src/bank/fee_distribution.rs [hunk 13]
```diff
@@ -50,24 +69,56 @@ impl Bank {
...
     }
 
+    // Deposits fees into a specified account and if successful, returns the new balance of that account
+    fn deposit_fees(
+        &self,
+        pubkey: &Pubkey,
+        fees: u64,
+        options: DepositFeeOptions,
+    ) -> Result<u64, DepositFeeError> {
+        let mut account = self.get_account_with_fixed_root(pubkey).unwrap_or_default();
```
Comment (t-nelson) [inline runtime/src/bank/fee_distribution.rs:98]: why not fail directly here rather than do the whole footgun-prone `unwrap_or_default()` thing?

**Review by t-nelson — COMMENTED** [review]
that the refactors have been squashed into the logical changes is pretty not great. this should have been at least two PRs, probably more.

have we checked how many mb validators will be effected by this change?

do we intend to prevent assignment of invalid fee-collection accounts?

File: runtime/src/bank.rs (hunk 12 above)
Comment (jstarry) [reply runtime/src/bank.rs:3809]: It's only used in tests right now... using the system program (default pubkey) as the highest staked node should also not happen outside of tests in any circumstances. I'll clean it up in a follow up change though to be sure

File: runtime/src/bank/fee_distribution.rs (hunk 13 above)
Comment (jstarry) [reply runtime/src/bank/fee_distribution.rs:98]: Because it's fine if the account doesn't exist as long as the fees top up the balance to at least rent exempt.

**Comment by jstarry** [comment]
> that the refactors have been squashed into the logical changes is pretty not great. this should have been at least two PRs, probably more.

Noted, will do that next time.

> have we checked how many mb validators will be effected by this change?

Right now the validator implementation forces nodes to use the fee collector account (also known as node id account) as the fee payer for their votes (therefore must already be a valid system account). And tpu votes also checks that the node id is the payer for the votes and uses that to weight vote transactions for block inclusion. So no one will be impacted.

> do we intend to prevent assignment of invalid fee-collection accounts?

Not now.. it isn't possible to enforce that a fee-collection account stays valid. It's owner can change at any time and it can be withdrawn to zero lamports at any time. Best thing I think we could do in the future is to create an enshrined "NodeAccount" or "FeeCollectorAccount" that must always be valid and used separately from fee payments. But that's a big lift and probably not necessary. If validators want fees, they won't create funky fee collection accounts.

**Comment by t-nelson** [comment]
this was _way_ too big to backport.  cc/ @willhickey

if we're developing a fix with intention to backport, it _must_ be done in the smallest change set possible. now we have a mass of refactor to sift through if we need to debug compatibility issues (which we are ATM)

File: runtime/src/bank.rs (hunk 12 above)
Comment (jstarry) [reply runtime/src/bank.rs:3809]: PR fix is being worked on here: https://github.com/solana-labs/solana/pull/34135
